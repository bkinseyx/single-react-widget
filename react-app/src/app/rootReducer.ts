import { combineReducers } from '@reduxjs/toolkit';

import counterReducer from '../features/reduxDemo/reduxDemoSlice';
import profileReducer from '../features/profileCard/profileSlice';
import toDoReducer from '../features/toDoCard/toDoSlice';
import userReducer from '../features/userCard/userSlice';

export const rootReducer = combineReducers({
  reduxDemo: counterReducer,
  profile: profileReducer,
  toDo: toDoReducer,
  user: userReducer,
});

export type RootState = ReturnType<typeof rootReducer>;

export default rootReducer;
