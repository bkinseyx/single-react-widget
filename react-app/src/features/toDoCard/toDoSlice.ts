import { createSlice } from '@reduxjs/toolkit';

import { createToDoReducer } from './reducers/createToDoReducer';
import { deleteToDoReducer } from './reducers/deleteToDoReducer';
import { updateToDoReducer } from './reducers/updateToDoReducer';
import { resetReducer } from './reducers/resetReducer';
import { setErrorMessageReducer } from './reducers/setErrorMessageReducer';

export interface ToDo {
  toDoId: number;
  description?: string;
  profileId: number;
  isComplete: boolean;
}

export interface ToDoState {
  toDoList: ToDo[];
  maxToDoId: number;
  errorMessage?: string;
}

export const initialState: ToDoState = {
  toDoList: [
    { toDoId: 1, profileId: 1, description: 'eat tacos', isComplete: true },
    { toDoId: 2, profileId: 1, description: 'drink milk', isComplete: false },
    {
      toDoId: 3,
      profileId: 1,
      description: 'walk and chew gum',
      isComplete: false,
    },
  ],
  maxToDoId: 3,
};

export const toDoSlice = createSlice({
  name: 'toDo',
  initialState,
  reducers: {
    createToDo: createToDoReducer,
    deleteToDo: deleteToDoReducer,
    updateToDo: updateToDoReducer,
    setErrorMessage: setErrorMessageReducer,
    reset: resetReducer,
  },
});

export const {
  createToDo,
  deleteToDo,
  updateToDo,
  setErrorMessage,
  reset,
} = toDoSlice.actions;

export default toDoSlice.reducer;
