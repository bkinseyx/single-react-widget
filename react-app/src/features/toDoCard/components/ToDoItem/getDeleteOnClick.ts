import { deleteToDo } from '../../toDoSlice';
import { store } from 'app/store';

export const getDeleteOnClick = (toDoId?: number) => (): void => {
  /* istanbul ignore next */
  if (!toDoId) {
    return;
  }
  store.dispatch(deleteToDo(toDoId));
};
