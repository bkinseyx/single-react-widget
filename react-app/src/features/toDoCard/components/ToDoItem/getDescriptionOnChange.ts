import { updateToDo } from '../../toDoSlice';
import { store } from 'app/store';

export const getDescriptionOnChange = (toDoId?: number) => (
  description: string,
): void => {
  /* istanbul ignore next */
  if (!toDoId) {
    return;
  }
  store.dispatch(
    updateToDo({
      toDoId,
      description,
    }),
  );
};
