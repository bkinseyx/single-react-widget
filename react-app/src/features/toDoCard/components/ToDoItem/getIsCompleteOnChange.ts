import { updateToDo } from '../../toDoSlice';
import { store } from 'app/store';

export const getIsCompleteOnChange = (toDoId?: number) => (
  isComplete: boolean,
): void => {
  /* istanbul ignore next */
  if (!toDoId) {
    return;
  }
  store.dispatch(
    updateToDo({
      toDoId,
      isComplete,
    }),
  );
};
