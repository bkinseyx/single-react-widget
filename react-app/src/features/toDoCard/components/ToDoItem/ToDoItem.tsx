import React from 'react';
import styles from './ToDoItem.module.css';

import { ToDo } from '../../toDoSlice';
import { CustomTextInput } from 'app/components/CustomTextInput/CustomTextInput';
import { emptyToDo } from '../../helpers/toDoTemplate';
import { getDescriptionOnChange } from './getDescriptionOnChange';
import { getDeleteOnClick } from './getDeleteOnClick';
import { getIsCompleteOnChange } from './getIsCompleteOnChange';
import { CustomCheckbox } from 'app/components/CustomCheckbox/CustomCheckbox';

interface ToDoProps {
  toDo: ToDo;
}

export const ToDoItem: React.FC<ToDoProps> = ({ toDo = emptyToDo }) => {
  return (
    <li className={'list-group-item ' + styles.toDoItem}>
      <CustomCheckbox
        idPrefix={`isComplete-${toDo.toDoId}`}
        label={'Complete?'}
        checked={toDo.isComplete}
        onChange={getIsCompleteOnChange(toDo.toDoId)}
      />
      <CustomTextInput
        idPrefix={`description-${toDo.toDoId}`}
        label={'Description'}
        value={toDo.description}
        onChange={getDescriptionOnChange(toDo.toDoId)}
      />
      <div className={styles.buttonWrapper}>
        <button
          className="btn btn-danger"
          onClick={getDeleteOnClick(toDo.toDoId)}
        >
          Delete
        </button>
      </div>
    </li>
  );
};
