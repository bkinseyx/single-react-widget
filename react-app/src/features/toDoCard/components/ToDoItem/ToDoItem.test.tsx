import React from 'react';
import {
  render,
  screen,
  fireEvent,
  RenderResult,
} from '@testing-library/react';
import { Provider } from 'react-redux';

import { store } from 'app/store';
import { ToDoItem } from './ToDoItem';
import { ToDo, reset } from 'features/toDoCard/toDoSlice';

const renderItem = (toDo: ToDo): RenderResult =>
  // the ToDoItem component is now "reduxified"
  render(
    <Provider store={store}>
      <ToDoItem toDo={toDo} />
    </Provider>,
  );

const getToDo = (toDoId: number): ToDo => {
  const toDo = store.getState().toDo.toDoList.find((p) => p.toDoId === toDoId);
  expect(toDo).not.toBeUndefined();
  // typescript doesn't understand that ToDo must be defined here, so we explicitly cast
  return toDo as ToDo;
};

test('renders ToDoItem', () => {
  const toDo = getToDo(1);
  const { asFragment } = renderItem(toDo);
  expect(asFragment()).toMatchSnapshot();
});

test('ToDoItem description field', () => {
  let toDo = getToDo(1);
  const { asFragment, rerender } = renderItem(toDo);
  let textInput = screen.getByLabelText('Description');
  expect(textInput).toHaveValue('eat tacos');
  fireEvent.change(textInput, { target: { value: 'live, laugh, love' } });

  textInput = screen.getByLabelText('Description');
  toDo = getToDo(1);
  expect(toDo).toMatchInlineSnapshot(`
    Object {
      "description": "live, laugh, love",
      "isComplete": true,
      "profileId": 1,
      "toDoId": 1,
    }
  `);

  // it is necessary to reduxify the component when you rerender also
  rerender(
    <Provider store={store}>
      <ToDoItem toDo={toDo} />
    </Provider>,
  );
  expect(textInput).toHaveValue('live, laugh, love');
  expect(asFragment()).toMatchSnapshot();
});

test('ToDoItem isComplete field', () => {
  // Whenever you do multiple redux tests in the same file, it's best practice to reset the state.
  // Otherwise the state from a previous test could bleed into a new one and cause unexpected issues.
  store.dispatch(reset());
  let toDo = getToDo(1);
  const { asFragment, rerender } = renderItem(toDo);

  // It's best practice to get an element the same way a real user would get an element.
  // In this case by the label text.
  const isCompleteCheckbox = screen.getByLabelText(/complete/i);
  expect(isCompleteCheckbox).toBeChecked();

  // this is how we check/uncheck a checkbox
  fireEvent.click(isCompleteCheckbox);
  toDo = getToDo(1);
  expect(toDo).toMatchInlineSnapshot(`
    Object {
      "description": "eat tacos",
      "isComplete": false,
      "profileId": 1,
      "toDoId": 1,
    }
  `);

  // it is necessary to reduxify the component when you rerender also
  rerender(
    <Provider store={store}>
      <ToDoItem toDo={toDo} />
    </Provider>,
  );
  expect(isCompleteCheckbox).not.toBeChecked();
  expect(asFragment()).toMatchSnapshot();
});
