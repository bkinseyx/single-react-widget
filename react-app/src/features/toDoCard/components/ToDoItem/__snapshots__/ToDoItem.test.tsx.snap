// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ToDoItem description field 2`] = `
<DocumentFragment>
  <li
    class="list-group-item toDoItem"
  >
    <div>
      <label
        for="isComplete-1-input"
      >
        Complete?
      </label>
      <input
        checked=""
        class="form-control checkbox"
        id="isComplete-1-input"
        type="checkbox"
      />
    </div>
    <div>
      <label
        for="description-1-input"
      >
        Description
      </label>
      <input
        class="form-control"
        id="description-1-input"
        type="text"
        value="live, laugh, love"
      />
    </div>
    <div
      class="buttonWrapper"
    >
      <button
        class="btn btn-danger"
      >
        Delete
      </button>
    </div>
  </li>
</DocumentFragment>
`;

exports[`ToDoItem isComplete field 2`] = `
<DocumentFragment>
  <li
    class="list-group-item toDoItem"
  >
    <div>
      <label
        for="isComplete-1-input"
      >
        Complete?
      </label>
      <input
        checked=""
        class="form-control checkbox"
        id="isComplete-1-input"
        type="checkbox"
      />
    </div>
    <div>
      <label
        for="description-1-input"
      >
        Description
      </label>
      <input
        class="form-control"
        id="description-1-input"
        type="text"
        value="eat tacos"
      />
    </div>
    <div
      class="buttonWrapper"
    >
      <button
        class="btn btn-danger"
      >
        Delete
      </button>
    </div>
  </li>
</DocumentFragment>
`;

exports[`renders ToDoItem 1`] = `
<DocumentFragment>
  <li
    class="list-group-item toDoItem"
  >
    <div>
      <label
        for="isComplete-1-input"
      >
        Complete?
      </label>
      <input
        checked=""
        class="form-control checkbox"
        id="isComplete-1-input"
        type="checkbox"
      />
    </div>
    <div>
      <label
        for="description-1-input"
      >
        Description
      </label>
      <input
        class="form-control"
        id="description-1-input"
        type="text"
        value="eat tacos"
      />
    </div>
    <div
      class="buttonWrapper"
    >
      <button
        class="btn btn-danger"
      >
        Delete
      </button>
    </div>
  </li>
</DocumentFragment>
`;
